Generate Excel analitics

mspiridonov2706 opened the pull request:
Задача:
https://www.notion.so/API-excel-d97e6256e3e244be91878445aa7583b8

Добавил роутер tasks и метод get_task_excel_report_request, который формирует excel файл в формате csv.

Commit:
feature: add task router

added report generation in task service;

Commit:
fix: spelling error

ADesBiysk reviewed (commented):
Прелагаю формировать отчёт на основе шаблона, чтобы это был законченный документ, а не заготовка, для которого нужно каждый раз настраивать оформление. Есть отличная библиотека https://openpyxl.readthedocs.io/en/stable/

mspiridonov2706 commented:
> 

а какое именно нужно оформление настроить?

mspiridonov2706 commented:
> Прелагаю формировать отчёт на основе шаблона, чтобы это был законченный документ, а не заготовка, для которого нужно каждый раз настраивать оформление. Есть отличная библиотека https://openpyxl.readthedocs.io/en/stable/

а какое именно нужно оформление настроить?

ADesBiysk commented:
> > Прелагаю формировать отчёт на основе шаблона, чтобы это был законченный документ, а не заготовка, для которого нужно каждый раз настраивать оформление. Есть отличная библиотека https://openpyxl.readthedocs.io/en/stable/
> 
> а какое именно нужно оформление настроить?

Предполагается, что результаты всех отчётов будут сохранятся в разных листах одного excel документа. Также csv формат не позволит использовать изменять размер шрифта, делать шрифт наклонным или полужирным, вставлять изображения, объединять ячейки и т.д.

mspiridonov2706 commented:
> > > Прелагаю формировать отчёт на основе шаблона, чтобы это был законченный документ, а не заготовка, для которого нужно каждый раз настраивать оформление. Есть отличная библиотека https://openpyxl.readthedocs.io/en/stable/
> > 
> > 
> > а какое именно нужно оформление настроить?
> 
> Предполагается, что результаты всех отчётов будут сохранятся в разных листах одного excel документа. Также csv формат не позволит использовать изменять размер шрифта, делать шрифт наклонным или полужирным, вставлять изображения, объединять ячейки и т.д.

Получается, когда пользователь обращается к роутеру, то он должен скачивать excel файл, который содержит все предыдущие отчёты плюс новый отчёт? Или он просто скачивает текущий отчёт, а объединять они уже сами будут?

ADesBiysk commented:
> > > > Прелагаю формировать отчёт на основе шаблона, чтобы это был законченный документ, а не заготовка, для которого нужно каждый раз настраивать оформление. Есть отличная библиотека https://openpyxl.readthedocs.io/en/stable/
> > > 
> > > 
> > > а какое именно нужно оформление настроить?
> > 
> > 
> > Предполагается, что результаты всех отчётов будут сохранятся в разных листах одного excel документа. Также csv формат не позволит использовать изменять размер шрифта, делать шрифт наклонным или полужирным, вставлять изображения, объединять ячейки и т.д.
> 
> Получается, когда пользователь обращается к роутеру, то он должен скачивать excel файл, который содержит все предыдущие отчёты плюс новый отчёт? Или он просто скачивает текущий отчёт, а объединять они уже сами будут?

Я бы делал так. Есть методы в сервисах, которые позволяют получать разные виды статистики. В этой задаче делается первый. 
Будут роутеры в которых будет передаваться эта информация на фронт (в процессе проектирования).
А также будет один роутер, который будет отдавать полный отчёт в формате xlsx. В этом файле на каждом листе будут располагаться разные типы отчётов. Каждый раз этот файл будет генерироваться заново. 
Файл генерируется следующим образом. Есть шаблон отчёта: нужное количество листов, на каждом есть область шапки, "табличной части" и подвала. Нужно заполнить данные в шапке. Сгенерировать нужное количество строчек "табличной части" и заполнить подвал. И так для каждого листа/отчёта. Результат отдать пользователю.
Формат csv очень удобен для генерации, но это не "конечный продукт", а всего лишь полуфабрикат. Сам часто его использую, для хранения измерительной информации с различных приборов. Чтобы привести его в вид удобный для работы пользователю нужно потратить много времени и делать это постоянно. Сформировать сразу готовый отчёт (см. отчёты в 1С) который не стыдно отправить другому человеку, распечатать или сравнить с предыдущими - не сложно. 
Твоя задача сделать образец, по которому потом будут делать другие отчёты.
Шаблона отчёта пока нет. Попробуй сделать сам. Если нужна будет помощь - напиши.

mspiridonov2706 commented:
> > > > > Прелагаю формировать отчёт на основе шаблона, чтобы это был законченный документ, а не заготовка, для которого нужно каждый раз настраивать оформление. Есть отличная библиотека https://openpyxl.readthedocs.io/en/stable/
> > > > 
> > > > 
> > > > а какое именно нужно оформление настроить?
> > > 
> > > 
> > > Предполагается, что результаты всех отчётов будут сохранятся в разных листах одного excel документа. Также csv формат не позволит использовать изменять размер шрифта, делать шрифт наклонным или полужирным, вставлять изображения, объединять ячейки и т.д.
> > 
> > 
> > Получается, когда пользователь обращается к роутеру, то он должен скачивать excel файл, который содержит все предыдущие отчёты плюс новый отчёт? Или он просто скачивает текущий отчёт, а объединять они уже сами будут?
> 
> Я бы делал так. Есть методы в сервисах, которые позволяют получать разные виды статистики. В этой задаче делается первый. Будут роутеры в которых будет передаваться эта информация на фронт (в процессе проектирования). А также будет один роутер, который будет отдавать полный отчёт в формате xlsx. В этом файле на каждом листе будут располагаться разные типы отчётов. Каждый раз этот файл будет генерироваться заново. Файл генерируется следующим образом. Есть шаблон отчёта: нужное количество листов, на каждом есть область шапки, "табличной части" и подвала. Нужно заполнить данные в шапке. Сгенерировать нужное количество строчек "табличной части" и заполнить подвал. И так для каждого листа/отчёта. Результат отдать пользователю. Формат csv очень удобен для генерации, но это не "конечный продукт", а всего лишь полуфабрикат. Сам часто его использую, для хранения измерительной информации с различных приборов. Чтобы привести его в вид удобный для работы пользователю нужно потратить много времени и делать это постоянно. Сформировать сразу готовый отчёт (см. отчёты в 1С) который не стыдно отправить другому человеку, распечатать или сравнить с предыдущими - не сложно. Твоя задача сделать образец, по которому потом будут делать другие отчёты. Шаблона отчёта пока нет. Попробуй сделать сам. Если нужна будет помощь - напиши.

Спасибо!
Не понял только один момент:
> Будут роутеры в которых будет передаваться эта информация на фронт (в процессе проектирования)

в данном случае информация на фронт передаётся в формате excel или в виде json строки? На счёт полного отчёта понял.

И ещё доп вопрос. Может тогда генерацию полного отчёта перенести в другой роутер? Сейчас это делается по ручке tasks/report и создаётся отчёт в сервисе task. Как-то нелогично выглядит.

Если я правильно понял логику то в сервисе Task должен остаться запрос статистики, который возвращается в виде json строки (или excel файл) и содержит информацию о том сколько каких заданий в общей сложности было принято/отклонено или не прислано. А полный отчёт перенести в отдельный роутер report например и по ручке excel-report/ грузить excel отчёт.

Commit:
fix: add openpyxl and template

add new router for generate full excel report.

mspiridonov2706 commented:
Очень надеюсь что понял Дениса правильно.

Реализовал следующую логику:

Есть файл шаблон в папке static report_template.xltx в котором заранее указаны все листы с нужным форматированием/header'ами и footer'ами.
Есть handler tasks/report который выдаёт excel отчёт по задачам, который содержит информацию о количестве сданных/несданных или отклонённых отчётов.

Есть handler full-excel-report/ который выдаёт не конкретный отчёт как в примере выше, а полный excel отчёт. Сейчас он выдаёт только информацию о задачах и тестовый лист, который я добавил для демонстрации.

Работает это с помощью класса ExcelReportService у которого реализованы следующие методы:

get_report_template() - возвращает workbook со всеми листами, который используется в методе generate_full_report() для генерации полного отчёта, либо workbook с одним листом для конкретного отчёта как в примере handler'a tasks/report. Реализовано это путём удаления лишних листов из workbook'а, так как копировать листы из одного файла в другой библиотека не позволяет. При этом листы удаляются именно в workbook, а не в самом файле шаблона. Поэтому он остаётся нетрунутым.

save_report_to_stream() - сохраняет итоговый workbook в байт-поток для дальнейшей передачи этого потока в StreamingResponse

generate_full_report() - сохраняет полный отчёт. Сейчас в него добавляются данные по task и тестовый лист для примера. В теории именно в этот метод в дальнейшем нужно добавлять вызовы методов для генерации полного отчёта.

create_tasks_statistics_report() - метод для создания отдельного отчёта по задачам. Потом таких методов может стать больше и их можно использовать для генерации как отдельных отчётов так и для генерации полного отчёта.

class Sheets - для упрощения получения различных листов из шаблона.

По идее в итоге всё должно работать таким образом:

Если нужно создать новый отчёт, то в классе ExcelReportService создаётся новый метод наподобие create_tasks_statistics_report() который берёт данные из базы и добавляет информацию в шаблон.

В методе generate_full_report() добавляется этот новый метод.

Создаётся новый handler, например shift/report, и в сервисе Shift прописывается новый метод для создания отчёта, который делается по подобию get_tasks_statistics_report()

Таким образом, как мне кажется, вполне неплохо реализован функционал как генерации полного отчёта так и отдельных маленьких отчётов

Commit:
fix: fix docstring

mspiridonov2706 commented:
<img width="1512" alt="Screen Shot 2022-12-29 at 01 28 01" src="https://user-images.githubusercontent.com/70363820/209880870-0ae7bfda-7b22-4e96-bf12-9f756e30de44.png">
<img width="1512" alt="Screen Shot 2022-12-29 at 01 27 51" src="https://user-images.githubusercontent.com/70363820/209880876-85c7ac71-a180-4d96-a59c-f7f830cfc9d2.png">
<img width="1512" alt="Screen Shot 2022-12-29 at 01 31 03" src="https://user-images.githubusercontent.com/70363820/209881016-0423f83b-43f6-402e-b18b-da0ecf7d36e8.png">

Получаются такие файлы:
1 скриншот - это файл с handler'a task/report в котором только 1 отчёт (1 лист внизу)
2 и 3 скриншот - это файл с handler'a full-excel-report/ в котором все отчёты (2 листа внизу Задачи и Тест)

ADesBiysk commented on pyproject.toml:
```diff
@@ -19,6 +19,7 @@ fastapi-restful = "^0.4.3"
 aiohttp = ">=3.8.3"
 aiolimiter = "1.0.0"
 pytz = "^2022.6"
+openpyxl = "^3.0.10"
```
Не забывай добавлять зависимости и в файл requirements.txt

ADesBiysk commented on src/core/services/excel_report_service.py:
```diff
@@ -0,0 +1,79 @@
+        """
+        workbook = load_workbook(settings.report_template_path)
+        workbook.template = False
+        if sheet_name:
```
Проблема с именами. В одной короткой функции есть: sheet_name, sheetnames, name.

ADesBiysk replied:
Такой способ не поможет если в шаблоне будет, например, 3 листа, а нужно будет сформировать только 2 отчёта. 
Логичнее удалять неиспользуемые листы уже после формирования всех отчётов.

mspiridonov2706 replied:
Хорошо. Об этом не подумал. Думал, что будет либо 1 отчёт, либо все сразу.

ADesBiysk commented on src/core/services/excel_report_service.py:
```diff
@@ -0,0 +1,79 @@
+    def __init__(self, task_repository: TaskRepository = Depends()) -> None:
+        self.__task_repository = task_repository
+
+    async def get_report_template(self, sheet_name: str | None = None) -> Workbook:
```
Насколько я помню, библиотека не асинхронная, поэтому делать эту функцию асинхронной нет смысла.

ADesBiysk commented on src/core/services/excel_report_service.py:
```diff
@@ -0,0 +1,79 @@
+        stream.seek(0)
+        return stream
+
+    async def generate_full_report(self) -> StreamingResponse:
```
Предлагаю передавать в виде аргумента список нужных отчётов. Тогда не потребуется делать несколько функций для генерации разного количества отчётов. Будет только одна, которая уже будет вызывать нужные приватные функции для формирования определённого отчёта.

ADesBiysk commented on src/core/services/excel_report_service.py:
```diff
@@ -0,0 +1,79 @@
+            sheet.cell(row=row_num, column=4, value=task.waiting)
+        return sheet
+
+    async def create_test_report(self, workbook: Workbook) -> Worksheet:
```
В ветке develop данная тестовая функция не нужна. Позже добавим другие статистики.

mspiridonov2706 replied:
понимаю, да. Я её перед вливанием собирался удалить. Сейчас чисто для примера оставил. Спасибо!

ADesBiysk commented on src/core/services/task_service.py:
```diff
@@ -24,3 +30,13 @@ async def get_task_by_day_of_month(self, tasks: Shift.tasks, day_of_month: int)
             raise TodayTaskNotFoundError()
         return task
+
+    async def get_tasks_statistics_report(self) -> StreamingResponse:
```
Давай сделаем так, чтобы другие сервисы не формировали сами excel файлы. За это будет отвечать только один сервис.

AntonZelinsky replied:
Согласен. Должно быть разделение кто что делает, этот сервис занимается получение данных для отчёта, передачей их в сервис для формирования отчёта и возврат результата клиену. Другой сервис должен иметь пару публичных функций, которые будут принимать данные для формирования отчётов, а дельше, где-то под капотом форматировать документ и заполнять данными.

ADesBiysk commented on src/api/routers/excel_report.py:
```diff
@@ -0,0 +1,26 @@
+
+from src.core.services.excel_report_service import ExcelReportService
+
+router = APIRouter(prefix="/full-excel-report", tags=["Excel"])
```
Неудачное название роутера.
Я писал про 2 роутера, имея в виду, что на некоторых страницах фронта может использоваться формируемая статистика, и нужно вынести формирование статистики в отдельные функции, которые уже можно будет использовать для формирования json или excel файла. Тебе нужно реализовать только формирование excel файла.

ADesBiysk replied:
Нужно оставить только один роутер. Тут возможно 2 крайних варианта:
1. Отталкиваемся от сущности, т.е. 
/tasks/analytics
/shift/{shift_id}/analytics
и т.д. Проблема только в общем отчёте, непонятно какой URI нужно будет использовать.
2. Делаем отдельный роутер:
/analytics/task
/analytics/shift/{shift_id}
Так совсем плохо. 
Предлагаю реализовать 1 вариант. Для общего отчёта что-нибудь придумаем позже.
Report у нас используется для другой сущности, поэтому предлагаю использовать analytics.

mspiridonov2706 replied:
@ADesBiysk 
идею понял, да. Я постарался так и сделать.
Сейчас роутер /tasks/report (переименую в /tasks/analytics) используется только для статистики по задачам, а роутер full-excel-report/ - для формирования общего отчёта. Мне название тоже не очень нравится, сделал быстро для того, чтобы показать саму идею.

Спасибо за комментарии!

Commit:
fix: change generate_report logic;

add openpyxl in requirements.txt;
change excel_router name to analytics_router;
change task's /report router to /analytics;

Commit:
Merge branch 'develop' into feature/report_excel

AntonZelinsky commented on src/application.py:
```diff
@@ -5,11 +5,13 @@
 from fastapi.staticfiles import StaticFiles
 
 from src.api.routers import (
```
Сделай, пожалуйста, импорт `routers`, а потом используй его для получения его роуетров `routers.report_router`

AntonZelinsky commented on src/api/routers/__init__.py:
```diff
@@ -1,10 +1,12 @@
-from src.api.routers.user import router as user_router  # noqa
+from src.api.routers.analytics import router as analytics_router  # noqa
+from src.api.routers.healthcheck import router as healthcheck_router  # noqa
+from src.api.routers.hello import (  # noqa
```
`hello`???

mspiridonov2706 replied:
Это не моё. В этом файле видимо не правились окончания строк.

AntonZelinsky commented on src/core/services/excel_report_service.py:
```diff
@@ -0,0 +1,73 @@
+    """Сервис для создания excel отчётов."""
+
+    def __init__(self, task_repository: TaskRepository = Depends()) -> None:
+        self.__task_repository = task_repository
```
Предлагаю, сделать отдельно класс генерирущий excel документ и в этот класс ничего не должно внедрятся — просто генератор. Пусть он имеет свои отдельные методы генерации различных отчётов, но никак не связан с репозиториями. Так, получится отделить генерацию отчётов от формирования данных. Тут нужно всегда думать в таком ключе, что если хочешь сделать слабую связанность, то нужно представить, что бы ты делал, если бы была необходимость делать генерацию сразу в несколько форматов, excel, csv, txt, json. Тут наверняка не делалось бы дублирование обращений к базе, а сами способы генерации подставлялись бы в зависимости от запроса. Вот нужно такого подхода и придерживаться, формировать данные и вызывать генерацию. 

Если тебе нужен отдельный сервис для получения общей аналитики, то создай его отдельно. Генератор эксель отчётов предлагаю так и назвать, это не сервис, это просто генератор, чтобы не было путаницы.

mspiridonov2706 replied:
Получается мне нужно сделать отдельно класс генератор excel который будет содержать в себе методы для обращения к шаблону, заполнение листов, удаление лишних листов и и создать отдельно сервис который будет всем этим управлять?

AntonZelinsky replied:
Да. Общий сервис аналитики, который вызывается из апи, он вызываем репозитоирй, который получает данные и данные передаёт в какой-то првайдер/билдер/генератор excel документов, который возвращает форматированный документ

AntonZelinsky commented on src/api/routers/analytics.py:
```diff
@@ -0,0 +1,27 @@
+        self,
+    ) -> StreamingResponse:
+        """Формирует excel файл со всеми отчётами."""
+        reports = tuple(report.value for report in ExcelReportService.Sheets)
+        return await self.excel_report_service.generate_report(reports)
```
Вроде метод специальный завёл в каком-то сервисе, а тут это, что-то вообще не понятно что, хорошо же дружили, не создавали бизнес логику во вьюшках.
![image](https://user-images.githubusercontent.com/1431279/211692199-8b95617b-d1ec-49d6-b4c4-c18c3fa2d420.png)

mspiridonov2706 replied:
Виноват. Не заметил что во вьюхе это сделал

AntonZelinsky commented on src/core/services/excel_report_service.py:
```diff
@@ -0,0 +1,73 @@
+        workbook = self.__get_report_template()
+        # создание отчётов
+        if self.Sheets.TASKS.value in reports:
+            await self.__create_tasks_statistics_report(workbook)
```
Лучше делай отдельные внешние методы, а всякие общие генерации делай приватными. Но я пока не разбирался в этом классе, после исправления этого ревью посмотрю.

Commit:
fix: add excel generator module

Commit:
fix: import router

ADesBiysk commented on src/excel_generator/tasks_excel_generator.py:
```diff
@@ -0,0 +1,42 @@
+        # создаём футер отчёта
+        footer_data = (
+            "ИТОГО:",
+            "=SUM(B2, B32)",
```
Задачи можно будет добавлять в БД и их общее количество изменится.

ADesBiysk commented on src/excel_generator/base.py:
```diff
@@ -0,0 +1,74 @@
+
+    def __init__(self) -> None:
+        self.workbook = Workbook()
+        self.workbook.remove_sheet(self.workbook.active)
```
Зачем нужно сразу удалять лист? 
И не нашёл, где загружается шаблон.

mspiridonov2706 replied:
У меня не получилось скопировать листы из шаблона по твоему предложенному варианту, поэтому я пошёл по пути заполнения данных в новом Workbook.
Удаление листа нужно потому что при создании нового Workbook создаётся сразу пустой лист, я его удаляю, чтобы был чистый Workbook и можно было начать создавать отчёт с любого листа без дополнительных проверок.

ADesBiysk commented on src/excel_generator/base.py:
```diff
@@ -0,0 +1,74 @@
+        TASKS = "Задачи"
+        TEST = "Тест"
+
+    class Styles:
```
(enum.Enum)?

ADesBiysk commented on src/excel_generator/base.py:
```diff
@@ -0,0 +1,74 @@
+        self.workbook = Workbook()
+        self.workbook.remove_sheet(self.workbook.active)
+
+    def clear_workbook(self):
```
Может не очищать, а создавать новый объект?

Commit:
Merge branch 'develop' of https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend into feature/report_excel

Commit:
Updated poetry

Commit:
Updated excel_generation

Commit:
Removed method clear_workbook

AlexeyPlz commented:
Ольга попросила подхватить эту задачу.

1. Обновил зависимости.
2. Исправил конфликты.
3. Убрал всё, что связано с тестовым генератором.
4. Количество строк в таблице теперь зависит от количества заданий.
5. Убрал и исправил импорты, теперь в сервисах импортируется только объект главного генератора.
6. Сделал часть методов приватными, так как используются только внутри своих классов.
7. Исправил tasks/analytics на tasks/analytics/, так как в противном случае FastAPI думал, что я перехожу по адресу tasks/<task_id> и выдавал ошибку по id.
8. Убрал метод очистки объекта workbook в ExcelBaseGenerator, теперь он просто пересоздаётся после сохранения отчёта в строке 21.

Возможно что-то забыл исправить, так как немного запутался в комментариях и изменениях, которые Иван уже сделал.

Также изучал вопрос шаблона, там Иван писал выше касательно этого.
Поэтому я пока не стал трогать, можно пока оставить так, чтобы фронтеры могли использовать уже.
А затем уже отдельной задачей займусь рефакторингом с использованием шаблона.

AntonZelinsky commented on src/api/routers/task.py:
```diff
@@ -73,3 +74,19 @@ async def get_task(
+        - cписок всех заданий;
+        - общее количество принятых/отклонённых/не предоставленных отчётов по каждому заданию.
+        """
+        return await self.task_service.generate_report()
```
Предлагаю вынести это из таск сервиса, да и вообще из роутера таск. Пусть у нас будет одна апи отвечающая за репорты, просто добалять в роут слово таскс и пусть всё будет там. 
Чтобы отделить основную логику приложения от аналитической
Метод в репозитории останется на своём месте, так как это конкретное получение данных конкретных таблиц.

AlexeyPlz replied:
@ADesBiysk
Предложение Антона можно реализовать двумя способами:
1) Один эндпоинт, где выбираем через query-параметр нужные данные;
2) Для каждого варианта свой эндпоинт.

Вопрос, какой лучше подойдёт?

AntonZelinsky replied:
2

AntonZelinsky commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,14 @@
+    def __init__(self, excel_report_service: ExcelReportService = Depends()) -> None:
+        self.__excel_report_service = excel_report_service
+
+    async def generate_report(self) -> StreamingResponse:
```
Если это полный репорт, то добавь инфомрацию об этом в название метода. Можно даже докстринги добавить

AntonZelinsky commented on src/excel_generator/main.py:
```diff
@@ -0,0 +1,10 @@
+    ...
+
+
+excel_generator = ExcelMainGenerator()
```
**Наследование — зло! Всегда предпочитайте ему агрегацию и композицию.**
Этот файл вообще что-то очень странное и непонятное. Что значит мэйн? Почему он наследуется от таск? А таск от бэйз...

Как же сильно не хватает знаний и структурности по ООП, SOLID и хотя бы базовым паттернам!:(((

AntonZelinsky commented on src/excel_generator/base.py:
```diff
@@ -0,0 +1,64 @@
+        sheet.column_dimensions["A"].width = self.Styles.WIDTH.value
+
+    class Sheets(str, enum.Enum):
+        TASKS = "Задачи"
```
Удалить от сюда. 
Базовый класс ничего не должен знать о потомках. Вижу 3 использования этого свойства, но ни разу не вижу причины для этого.
Есть класс генерирующий отчёт для задач, там достаточно сделать переменную с названием книги и использовать при создании её. Всё! Зачем ещё использовать, я не понимаю

AntonZelinsky commented on src/excel_generator/tasks_excel_generator.py:
```diff
@@ -0,0 +1,33 @@
+    """Генератор excel отчёта по задачам."""
+
+    async def generate_tasks_statistics_report(self, data: tuple[TasksExcelReportDto]) -> Worksheet:
+        sheet: Worksheet = self.workbook.create_sheet(self.Sheets.TASKS.value)
```
Если `workbook` берётся из класса родителя, то должна быть защищённой, а не публичной.
Так же, все методы базового класса тоже должны быть защищёнными.

Но вообще, я предлагаю в конструктор этого класса передавать ExcelBaseGenerator и потом вызывать его методы для формирования документа. То есть, один занимается формированием документа и ему всё равно что в него передают, а другой управляет им. И таких управляющих может быть много, для разных типов отчётов. Классический паттерн [билдер](https://refactoring.guru/ru/design-patterns/builder). И как бонус, всё будет построено на агрегации и получится избавиться от наследования

Commit:
Changed generator logic

Commit:
Merge branch 'develop' of https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend into feature/report_excel

AntonZelinsky commented on src/core/settings.py:
```diff
@@ -66,6 +66,11 @@ def registration_template_url(self) -> str:
+    @property
+    def report_template_path(self) -> str:
+        """Путь до excel шаблона для формирования отчёта."""
+        return BASE_DIR / "src" / "static" / "report_template.xltx"
```
Что-то не вижу файла в закомичннных файлах. Как и использования этой переменной, может какой-то файл с использованием я скрыл, прошу указать

ADesBiysk commented on src/api/routers/analytics.py:
```diff
@@ -0,0 +1,42 @@
+    analitics_service: AnaliticsService = Depends()
+
+    @router.get(
+        "/",
```
Не совсем логично, что если запрашивается / то получаем полный отчёт. Я бы добавил total для однозначности.

ADesBiysk commented on src/api/routers/analytics.py:
```diff
@@ -0,0 +1,42 @@
+    )
+    async def generate_tasks_report(self) -> StreamingResponse:
+        """
+        Формирует excel файл со сведениями о задачах по всем сменам.
```
"Формирует отчёт с общей статистикой выполнения задач во всех сменах".
Упоминание Excel файла здесь точно не нужно.

ADesBiysk commented on src/excel_generator/base.py:
```diff
@@ -0,0 +1,64 @@
+from openpyxl.worksheet.worksheet import Worksheet
+
+
+class ExcelBaseGenerator:
```
Плохо, что Report уже занято и использовать его в аналитике не очень хорошо. Как вариант, можно использовать сочетание AnalyticReport.
Если следовать шаблону "Строитель", то
1. Должен быть интерфейс для строителя отчётов, например AnalyticReportBuilder.
2. Этот класс ExcelBaseGenerator должен быть ExcelAnalyticReportBuilder и наследоваться от него.

ADesBiysk replied:
А в этом классе будет приватный атрибут - формируемый документ, и метод для его получения.

ADesBiysk commented on src/excel_generator/tasks_excel_generator.py:
```diff
@@ -0,0 +1,38 @@
+from src.excel_generator.base import ExcelBaseGenerator
+
+
+class ExcelTasksGenerator:
```
Это будет AnalyticReportDirector, которому передаётся уже готовый объект AnalyticReportBuilder. Не через Depends().

ADesBiysk commented on src/excel_generator/tasks_excel_generator.py:
```diff
@@ -0,0 +1,38 @@
+        self.sheet_name = "Задачи"
+
+    async def generate_tasks_statistics_report(self, data: tuple[TasksExcelReportDto]) -> Worksheet:
+        sheet = self.__base_generator.create_sheet(self.sheet_name)
```
Этот класс не должен знать про Worksheet. Формируемый документ хранится в AnalyticReportBuilder.

ADesBiysk commented on src/excel_generator/tasks_excel_generator.py:
```diff
@@ -0,0 +1,38 @@
+            "Кол-во отклонённых отчётов",
+            "Кол-во не предоставленных отчётов",
+        )
+        self.__base_generator.fill_row(sheet, header_data, 1)
```
Так же надо убрать отсюда номер строки. Пусть их считает Builder, и тогда функцию правильнее будет назвать add_row(row_data). Лист и номер строки не нужно будет передавать.

ADesBiysk commented on src/excel_generator/tasks_excel_generator.py:
```diff
@@ -0,0 +1,38 @@
+            self.__base_generator.fill_row(sheet, astuple(task), row_num)
+        footer_data = (
+            "ИТОГО:",
+            f"=SUM(B2, B{total_tasks})",
```
Если посчитать и передать результаты, то полностью отвяжемся от формата в котором сохраняется отчёт.

Commit:
Merge branch 'develop' of https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend into feature/report_excel

AntonZelinsky commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,24 @@
+
+    async def generate_tasks_report(self) -> StreamingResponse:
+        """Репорт с задачами со всех смен."""
+        return await self.__excel_report_service.generate_report((self.__tasks_excel_generator.sheet_name,))
```
Ничем не отличается от того что выше. `generate_task_report`

AntonZelinsky commented on src/core/services/excel_report_service.py:
```diff
@@ -0,0 +1,38 @@
+        return StreamingResponse(stream, headers=headers)
+
+    async def generate_report(self, reports: tuple[str]) -> StreamingResponse:
+        if self.__tasks_excel_generator.sheet_name in reports:
+            await self.__generate_tasks_statistics_report()
+        return await self.__get_excel_report_response()
```
Должен возвращаться что-то сгенерированае, сами классы не хранят состояния.

AntonZelinsky commented on src/excel_generator/base.py:
```diff
@@ -0,0 +1,64 @@
+
+    def __init__(self) -> None:
+        self.__workbook = Workbook()
+        self.__workbook.remove_sheet(self.__workbook.active)
```
Как писал выше, это тут не храним, передаём в каждую функцию книгу

AntonZelinsky reviewed (commented):
Глянь как делали реализацию на этом проекте https://github.com/Studio-Yandex-Practicum/Lubimovka_backend/blob/develop/apps/feedback/services/spreadsheets.py
Просто для расширения кругозора, может что-то приглянется

Commit:
Changed service, generator for excel

Commit:
Changed checking for add_last_row in TaskBuilder

AlexeyPlz commented:
Переделал сервис и генератор (теперь это Builder и Director).
Классы строителя и дерактора не знаю о состоянии Workbook, как говорил Антон, он создаётся в сервисе и затем передаётся по функциям директору, который уже при помощи строителей создаёт лист с данными.
Worksheet теперь является аргументом у строителей, но по-умолчанию он None, так как строителю необходимо передать сначала Workbook, от которого он создаст Worksheet. То есть Workbook можно считать фундаментом, который нужно дать строителю.

AntonZelinsky commented on src/core/services/analytic_report_service.py:
```diff
@@ -0,0 +1,53 @@
+        """Генерация отчёта с заданиями."""
+        data = await self.__task_repository.get_tasks_statistics_report()
+        builder = AnalyticTaskReportBuilder()
+        return await AnalyticReportDirector(builder).generate_report(data, workbook)
```
Почему не через агрегацию или хотябы композицию (через конструктор) создаются экземпляры класса?

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,79 @@
+    """Интерфейс строителя."""
+
+    def __init__(self, sheet_name: str, header_data: tuple[str], footer_data: tuple[str]) -> None:
+        self.sheet_name = sheet_name
```
Ох и пришлось поломать голову по поводу этого конструктора чтобы понять кто и зачем его вызывает.

приватные поля должны быть приватными!!! Но они здесь вообще не нужны, опять же, этот файл не должен иметь своего состояния и этот же экземпляр могут использовать сразу для генераций нескольких документов, в текущей реализации это не возможно.
Удалить все переменные этого класса. Не стесняйся передавать их по ссылке в аргументы функции.

AntonZelinsky commented on src/core/services/analytic_report_service.py:
```diff
@@ -0,0 +1,53 @@
+        headers = {'Content-Disposition': f'attachment; filename={filename}'}
+        return StreamingResponse(stream, headers=headers)
+
+    async def generate_workbook(self) -> Workbook:
```
`generate_workbook`, `__get_report_response`, `__save_report` выносим в файл билдера

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,79 @@
+        """Создаёт лист внутри отчёта."""
+        self.sheet = workbook.create_sheet(self.sheet_name)
+
+    def add_first_row(self) -> None:
```
add_header

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,79 @@
+            row_number = index + 2
+            self.__add_row(astuple(task), row_number)
+
+    def add_last_row(self) -> None:
```
add_footer

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,79 @@
+        """Заполняет последнюю строку в листе."""
+        self.__add_row(self.footer_data, self.data_count + 2)
+
+    def make_styling(self):
```
apply_styles

AntonZelinsky commented on src/core/services/analytic_report_service.py:
```diff
@@ -0,0 +1,53 @@
+
+    async def __generate_task_report(self, workbook: Workbook) -> Workbook:
+        """Генерация отчёта с заданиями."""
+        data = await self.__task_repository.get_tasks_statistics_report()
```
Давай эту функцию вынесем расформируем и пренесём вызов  в методы генерации отчётов перенести в AnaliticsService. Так, избавимся от лишнего класса. Может как-то код и обобщим, но сначало хотелось бы к ± хорошей структуре привести.

AntonZelinsky reviewed (commented):
Не последняя итерация, но осталось чуть-чуть.)

Commit:
Removed __init__ for builder interface and report_service

Commit:
Updated task_builder fields to private

Commit:
Added noqa R504 for analytics_service

ADesBiysk commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,32 @@
+
+    def add_footer(self) -> None:
+        if self._data_count > 31:
+            footer_data = [
```
Эти данные в инициализаторе дублируются.

ADesBiysk commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,32 @@
+        self._data_count: int = 0
+
+    def add_footer(self) -> None:
+        if self._data_count > 31:
```
Лучше не использовать "магических" чисел. Возможно, заказчик захочет удалить часть заданий.

ADesBiysk commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,94 @@
+        """Заполняет строку данными из БД."""
+        self.__set_data_count(data)
+        for index, task in enumerate(data):
+            row_number = index + 2
```
А если "шапка" будет больше чем одна строка, то всё сломается.
Предлагаю при добавлении ряда (в __add_row) просто инкрементировать переменную, хранящую номер ряда. Тогда и __set_data_count не понадобится.
В специфических случаях можно её самостоятельно менять.

ADesBiysk commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,94 @@
+                cell.font = self.Styles.FONT_STANDART.value
+                cell.alignment = self.Styles.ALIGNMENT_STANDART.value
+        # задаём стиль для футера
+        footer_cells = list(self._worksheet.iter_rows())[max_row - 1]
```
Так будет работать [-1]?

ADesBiysk commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,94 @@
+        """Задаёт форматирование отчёта."""
+        max_row = self._data_count + 2
+        # задаём стиль для хэдера
+        header_cells = list(self._worksheet.iter_rows())[0]
```
Нужно сформировать список один раз и уже брать от него срезы. Формировать список 4 раза, чтобы в двух случаях взять только одно значение - очень расточительно.

ADesBiysk commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,94 @@
+        """Заполняет первую строку в листе."""
+        self.__add_row(self._header_data, 1)
+
+    def add_data(self, data: tuple[str]) -> None:
```
Аннотация неправильная.

ADesBiysk commented on src/excel_generator/analytic_report_director.py:
```diff
@@ -0,0 +1,27 @@
+        self, data: tuple, last_sheet: bool, workbook: Optional[Workbook] = None
+    ) -> Workbook | StreamingResponse:
+        """Генерация листа с данными."""
+        workbook = await self.__builder.create_workbook() if not workbook else workbook
```
Не понимаю, зачем в этом методе нужен Workbook. Так мы привязываемся к типу возвращаемого результата. Т.е. если понадобится сохранить в другой формат, то надо будет описывать новый класс. От этого нельзя наследоваться, поскольку будет нарушен принцип SOLID - L.
Мы же получаем в качестве параметра класс AnalyticReportBuilder, вот из него и можно забрать результат. Т.е. сделать так, чтобы директор только формировал последовательность шагов, но не возвращал результат.

AlexeyPlz replied:
> Не понимаю, зачем в этом методе нужен Workbook. Так мы привязываемся к типу возвращаемого результата. Т.е. если понадобится сохранить в другой формат, то надо будет описывать новый класс. От этого нельзя наследоваться, поскольку будет нарушен принцип SOLID - L. Мы же получаем в качестве параметра класс AnalyticReportBuilder, вот из него и можно забрать результат. Т.е. сделать так, чтобы директор только формировал последовательность шагов, но не возвращал результат.

Антон сказал workbook должен создаваться в начале и передаваться постоянно.

ADesBiysk replied:
Я ориентируюсь на эту схему:
![2023-02-09_092829](https://user-images.githubusercontent.com/26863439/217700033-b75b2ace-deea-4ca8-9e01-169c7f4b9ec7.png)
Ждём комментарии Антона. Мне самому интересно разобраться и узнать где я ошибаюсь.

Ivan-Skvortsov replied:
Метод [create_workbook ](https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend/blob/75d2c4f404f1e82e49ca280b1c3303590d2b17fd/src/excel_generator/builder.py#L36)не асинхронный. 

И да, на мой взгляд манипуляции с workbook тоже выглядят странновато)

ADesBiysk replied:
@AlexeyPlz Алексей, немного обсудили данный PR с Антоном. Он хотел убрать промежуточный объект Директор. Поскольку у нас не очень много отчётов и форматов, то все действия по формированию отчёта можно выполнять прямо в AnalyticsService.
 Если сейчас посмотреть, то действительно, метод AnalyticsService.__generate_task_report получает данные и дёргает директора. Т.е. этот класс является, по сути, промежуточным.

Давай удалим директора и перенесём вызов методов Строителя в сервис. Только объект workbook создаёт строитель и мы забираем уже полностью готовый результат (файл).
И мне кажется, что можно обойтись без параметра last_sheet.

ADesBiysk commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,32 @@
+        self._worksheet: Optional[Worksheet] = None
+        self._data_count: int = 0
+
+    def add_footer(self) -> None:
```
Эта функция, судя по всему, не нужна.

ADesBiysk commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,32 @@
+    """Строитель отчёта для заданий."""
+
+    def __init__(self) -> None:
+        self._sheet_name: str = "Задачи"
```
Всё-таки эти атрибуты нужно добавить в родительский класс, а здесь переопределять.

AlexeyPlz replied:
> Всё-таки эти атрибуты нужно добавить в родительский класс, а здесь переопределять.

Я так и сделал, но Антон сказал надо наоборот убрать, вы с ним противоположности иногда говорите)

ADesBiysk commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,34 @@
+
+    async def generate_full_report(self) -> StreamingResponse:
+        """Генерация полного отчёта."""
+        workbook = await self.__generate_task_report(last_sheet=True)
```
И если результат забирать из билдера, то и параметр last_sheet не понадобится.

AntonZelinsky replied:
Предлагаю этот всё чуть-чуть переделать (согласен, что идея с last_sheet очень плохая(. 
Я предлагаю в функциях генерации отчёта вызывать создание воркбука и получать пустой воркбук (как после вызова функции create_workbook), потом, передавать его вместе с конфигурационным дата классом (данных из бд для заполнения талицы должны заполняться в датакласс тут) в функцию генерации отчётов и там уже будет происходить создание страницы и заполнение её данными. Назад ничего возвращать не надо, так как объекты передаются по ссылке.

И того, получится так, что когда вызывается генерация отчёта, сначала достаются данные из бд, заполняется дата класс, создаётся книга и передаётся вместе с дата классом в метод generate_report. Там происходит создание страницы и заполнение данными. Потом, эта функция(generate_full_report) будет генерировать стрим и отдавать его тому кто вызвал её.
Всё.

Сейчас уже есть две апишки, в одной пусть будет просто генерация отчёта, во второй два раза (два экземпляра дата класса) генерация отчёта, так как сейчас если сделать это, то будет кривой excel на второй странице и стили почему-то не применяются.

Commit:
Changed builders

Commit:
Fix

ADesBiysk commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,91 @@
+        headers = {'Content-Disposition': f'attachment; filename={filename}'}
+        return StreamingResponse(stream, headers=headers)
+
+    async def create_workbook(self) -> Workbook:
```
Нет смысла делать асинхронной.

ADesBiysk commented on src/api/routers/analytics.py:
```diff
@@ -0,0 +1,42 @@
+    )
+    async def generate_task_report(self) -> StreamingResponse:
+        """
+        Формирует отчёт с общей статистикой выполнения задач во всех сменахм.
```
сменахМ.

ADesBiysk commented on src/api/routers/analytics.py:
```diff
@@ -0,0 +1,42 @@
+
+@cbv(router)
+class AnalyticsCBV:
+    analitics_service: AnaliticsService = Depends()
```
analYtics_service: AnalYticsService

ADesBiysk commented on src/core/db/DTO_models.py:
```diff
@@ -21,6 +21,14 @@ class FullReportDto:
 
 
+@dataclass
+class TasksExcelReportDto:
```
TasksAnalyticReportDto

ADesBiysk commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,34 @@
+from src.excel_generator.task_builder import AnalyticTaskReportBuilder
+
+
+class AnaliticsService:
```
AnalYticsService

ADesBiysk commented on src/excel_generator/analytic_report_director.py:
```diff
@@ -0,0 +1,27 @@
+        self.__builder = builder
+
+    async def generate_report(
+        self, data: tuple[str | int], last_sheet: bool, workbook: Optional[Workbook] = None
```
Неправильные аннотации типов. Д.б. tuple[TasksExcelReportDto]

ADesBiysk commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,91 @@
+        for data in self._header_data:
+            self.__add_row(data)
+
+    def add_data(self, data: tuple[tuple[str | int]]) -> None:
```
Неправильные аннотации типов. Д.б. tuple[TasksExcelReportDto]

ADesBiysk commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,34 @@
+
+    async def __generate_task_report(self, last_sheet: bool, workbook: Optional[Workbook] = None) -> Workbook:
+        """Генерация отчёта с заданиями."""
+        data = await self.__task_repository.get_tasks_statistics_report()
```
data - не очень информативное имя.

ADesBiysk commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,91 @@
+            cell.font = self.Styles.FONT_BOLD.value
+            cell.alignment = self.Styles.ALIGNMENT_HEADER.value
+        # задаём стиль для ячеек с данными
+        data_rows = rows[1:self._row_count - 1]  # fmt: skip
```
```suggestion
        data_rows = rows[1:-1] 

```

ADesBiysk commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,27 @@
+        self._row_count: int = 0
+
+    def add_footer(self) -> None:
+        self._footer_data = [
```
Лучше использовать один тип данных, а то сейчас все данные в кортежах хранятся, а эти в списке. И аннотации типов надо исправить выше.

AntonZelinsky replied:
Опечатка

AntonZelinsky replied:
Опечатка

ADesBiysk commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,91 @@
+
+    def __add_row(self, data: tuple[str | int]) -> None:
+        self._row_count += 1
+        for index, value in enumerate(data):
+            self._worksheet.cell(row=self._row_count, column=index + 1, value=value)
```
```suggestion
        for index, value in enumerate(data, start=1):
            self._worksheet.cell(row=self._row_count, column=index, value=value)

```

Commit:
Merge branch 'develop' of https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend into feature/report_excel

Commit:
Some changes for Analytics

ADesBiysk commented:
Алексей, немного обсудили данный PR с Антоном. Он хотел убрать промежуточный объект Директор. Поскольку у нас не очень много отчётов и форматов, то все действия по формированию отчёта можно выполнять прямо в AnalyticsService.
Если сейчас посмотреть, то действительно, метод AnalyticsService.__generate_task_report получает данные и дёргает директора. Т.е. этот класс является, по сути, промежуточным.

Давай удалим директора и перенесём вызов методов Строителя в сервис. Только объект workbook создаёт строитель и мы забираем уже полностью готовый результат (файл).
И мне кажется, что можно обойтись без параметра last_sheet.

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,93 @@
+    def add_header(self) -> None:
+        """Заполняет первые строки в листе."""
+        for data in self._header_data:
+            self.__add_row(data)
```
Кажется, что заголовок с колонками — это всего одна строка. Тут не нужен цикл.

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,93 @@
+
+    def add_footer(self) -> None:
+        """Заполняет последнюю строку в листе."""
+        self.__add_row(self._footer_data)
```
Если тут просто вызывается добавление строки, то нет смысла от копирования значения футера в переменную класса, а потом использования в функции с добавлением ещё одной строки.
 Просто тут сделать виртуальный метод, а в наследниках перегружать его и добавлять все значения

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,93 @@
+        stream = await self.__save_report(workbook)
+        filename = f"report_{datetime.now()}.xlsx"
+        headers = {'Content-Disposition': f'attachment; filename={filename}'}
+        return StreamingResponse(stream, headers=headers)
```
Эта функция должна возвращать только стрим, респонс должен формировать уровень представления. Например, если я захочу через бот получать отчёты в виде excel, то нужно будет сделать ещё одну функцию, либо перегружать эту, в общем, куча проблем.

А возвращаться в generate_report сразу два разных типа ответа — это зло! Показатель того, что кто-то не знает принципы SOLID и проблемы с проектированием

Вообще, можно подумать над тем, чтобы сделать метод публичным статическим, принимающим воркбук и возвращающим потом, чтобы отвзятаться от функции генерации. Либо сделать ещё один метод в классе AnalyticReportDirector и вызывать после генерации основного отчета.

AntonZelinsky commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,28 @@
+    """Строитель отчёта для заданий."""
+
+    def __init__(self) -> None:
+        self._sheet_name: str = "Задачи"
+        self._header_data: tuple[tuple[str]] = (
+            ("Задача", "Кол-во принятых отчётов", "Кол-во отклонённых отчётов", "Кол-во не предоставленных отчётов"),
+        )
+        self._footer_data: Optional[tuple[str]] = None
+        self._worksheet: Optional[Worksheet] = None
+        self._row_count: int = 0
```
Эту всю информацию нужно хранить в отдельном месте, я предлагаю, создать дата класс, который будет в себе хранить только данные необходимые для генерации определённого типа отчёта.
Нужно создать базовый дата класс только с определением типов, а потом от него делать наследники с установкой значений необходимых для генерации данных.
Этот класс нужно будет передавать в условную функцию generate_report, которая будет передавать его в функции выполняющие необходимые действия. Так же, в эти функции должны передаваться воркбуки или воркшит (они не должны храниться в дата классе, тут там только конфигурация отдельного отчёта).

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,93 @@
+
+    def create_sheet(self, workbook: Workbook) -> None:
+        """Создаёт лист внутри отчёта."""
+        self._worksheet = workbook.create_sheet(self._sheet_name)
```
этот класс не должен хранить состояние, воркбуки или воркшиты всегда передаются в аргументах методов.

vatut007 committed:
Merge remote-tracking branch 'origin/develop' into feature/report_excel

vatut007 committed:
Р”РѕРїРѕР»РЅРёР» Р·Р°РґР°С‡Сѓ. РЎРґРµР»Р°Р» РґР°С‚Р° РєР»Р°СЃСЃС‹

РР·РјРµРЅРёР» СЂР°Р±РѕС‚Сѓ РіРµРЅРµСЂР°С†РёРё РѕС‚С‡С‘С‚РѕРІ

ADesBiysk commented on src/api/routers/analytics.py:
```diff
@@ -0,0 +1,42 @@
+        self,
+    ) -> StreamingResponse:
+        """Формирует excel файл со всеми отчётами."""
+        return await self.analytics_service.generate_full_report()
```
Потом нужно будет добавить проверку токенов.

vatut007 replied:
Здесь не понял что нужно сделать

ADesBiysk replied:
Когда всё будет работать, нужно проверять наличие "живого" токена.

ADesBiysk commented on src/core/db/DTO_models.py:
```diff
@@ -32,6 +34,14 @@ class FullReportDto:
+    description: str
+    approved: int
+    declined: int
+    waiting: int
```
Нужно заменить статус waiting на skipped (пропущенные), поскольку изменилась логика изменения статусов.

ADesBiysk commented on src/core/db/migrations/versions/2023-01-18_12.25.57_added_default_tasks.py:
```diff
@@ -22,7 +22,7 @@
 def upgrade():
     # ### commands auto generated by Alembic - please adjust! ###
-    for task in json.load(open('tasks.json', 'r')):
+    for task in json.load(open('tasks.json', 'r', encoding='utf-8')):
```
Это будет исправлено в другой задаче для новых студентов.

ADesBiysk commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,55 @@
+
+    async def generate_full_report(self) -> StreamingResponse:
+        """Генерация полного отчёта."""
+        workbook = load_workbook(FILE_PATH)
```
Если первым вызвать этот метод, то будет вызвано исключение FileNotFoundError. Этого файла ещё нет.

ADesBiysk commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,55 @@
+    async def generate_task_report(self) -> StreamingResponse:
+        """Генерация отчёта с заданиями."""
+        workbook = self.__task_report_builder.create_workbook()
+        tasks = await self.__task_repository.get_tasks_statistics_report()
```
tasks не очень удачное название. Поэтому далее tasks_statistic=tasks выглядит не очень логичным.

AntonZelinsky replied:
А почему вообще не так делается? На э то о есть какие-то причины?
```
        workbook = await self._analytic_report_builder.get_report_response(workbook)
```

AntonZelinsky replied:
И кстати, возвращаешь ты не воркбук, а респонс или стрим. Название переменной не корректно

ADesBiysk commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,55 @@
+        workbook = self.__task_report_builder.create_workbook()
+        tasks = await self.__task_repository.get_tasks_statistics_report()
+        workbook = await self.__generate_task_report(tasks_statistic=tasks, workbook=workbook)
+        if os.path.exists(FILE_PATH) is False:
```
Лучше использовать библиотеку pathlib.

ADesBiysk commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,55 @@
+            WORKBOOK_FULL_DIR.mkdir(exist_ok=True)
+            workbook.save(FILE_PATH)
+        else:
+            workbook_full = load_workbook(FILE_PATH)
```
Здесь непонятная логика. Если директория есть, то загружаем файл и заново формируем лист. В результате выполняем генерацию 2 раза и если запустить формирование отчёта несколько раз, то в итоговом документе будет несколько листов. 
Наличие директории не нужно каждый раз проверять. Нужно один раз её создать при старте приложения.

ADesBiysk commented on src/core/services/task_service.py:
```diff
@@ -21,9 +21,7 @@ async def __download_file(self, file: UploadFile) -> str:
-    async def get_task_ids_list(
-        self,
-    ) -> list[UUID]:
+    async def get_task_ids_list(self) -> list[UUID]:
```
Такие изменения лучше убирать. Они только отвлекают. Если это изменяет автоматически  black, то пусть остаются.

ADesBiysk commented on src/excel_generator/analytic_report_director.py:
```diff
@@ -0,0 +1,34 @@
+        self.__builder = builder
+
+    async def generate_report(
+        self, data: tuple[TasksAnalyticReportDto],
```
self на отдельной строке.

ADesBiysk commented on src/excel_generator/analytic_report_director.py:
```diff
@@ -0,0 +1,34 @@
+        workbook: Optional[Workbook] = None,
+        worksheet: Optional[Worksheet] = None,
+        analytic_task_report_full: Optional[AnalyticTaskReportFull] = None
+    ) -> Workbook | StreamingResponse:
```
В каком случае вернётся StreamResponse?

ADesBiysk commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,108 @@
+    """Интерфейс строителя."""     
+    def __add_row(self, analytic_task_report: Optional[AnalyticTaskReportFull],
+                  data: tuple[str | int], worksheet: Optional[Worksheet]) -> None:
+        analytic_task_report.row_count += 1
```
Надо поправить аннотации типов. Если AnalyticTaskReportFull можно не передавать, то метод на первой строчке вызовет исключение. И это общий строитель для всех отчётов, поэтому нужно указать базовый класс, а не наследника.

ADesBiysk replied:
И назвать её можно просто ReportSettings.

ADesBiysk commented on src/excel_generator/analytic_report_director.py:
```diff
@@ -0,0 +1,34 @@
+        analytic_task_report_full: Optional[AnalyticTaskReportFull] = None
+    ) -> Workbook | StreamingResponse:
+        """Генерация листа с данными."""
+        worksheet = self.__builder.create_sheet(
```
Функция возвращает workbook.

ADesBiysk commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,55 @@
+        """Генерация отчёта с заданиями."""
+        return await AnalyticReportDirector(self.__task_report_builder).generate_report(
+            data=tasks_statistic, workbook=workbook,
+            analytic_task_report_full=AnalyticTaskReportFull, worksheet=Worksheet
```
Нужно передать экземпляры классов AnalyticTaskReportFull и Worksheet или сам класс?

vatut007 replied:
Нужно передать класс AnalyticTaskReportFull, data classs с данными из базы данных и workbook. Worksheet получается не нужно передавать. Я его уберу в следущем коммите

ADesBiysk commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,33 @@
+
+    @classproperty
+    def footer_data(self):
+        result = [
```
Нужно сразу формировать кортеж, а то сейчас сначала формируется список, и потом уже преобразуется в кортеж.

ADesBiysk commented on src/core/db/DTO_models.py:
```diff
@@ -69,3 +79,11 @@ class ShiftByUserWithReportSummaryDto:
+
+
+@dataclass
+class AnalyticTaskReportSheetDTO:
```
Это не DTO модель и ей здесь не место.

ADesBiysk commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,33 @@
+        return self.fget(owner_cls)
+
+
+class AnalyticTaskReportFull(AnalyticTaskReportSheetDTO):
```
Название совсем не понятное. Может TaskAnalyticReportSettings?
А базовый - BaseAnalyticReportSettings.

ADesBiysk commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,33 @@
+
+class AnalyticTaskReportFull(AnalyticTaskReportSheetDTO):
+    """Строитель отчёта для заданий"""
+    def __init__(self):
```
А зачем так делать?

AntonZelinsky replied:
Согласен, писал уже об этом раньше, у нас одна строка хидеров, а не две, следовательно, не зачем делать что-то вне задачи.
tuple[str]
А заодно и от цикла избавимся в другой функции

ADesBiysk reviewed (commented):
Замечаний много, но они небольшие.

vatut007 committed:
Исправил замечания

ADesBiysk commented on src/core/db/repository/task_repository.py:
```diff
@@ -19,3 +20,23 @@ async def get_task_ids_list(self) -> list[UUID]:
+                Task.description,
+                func.count().filter(Report.status == Report.Status.APPROVED).label(Report.Status.APPROVED),
+                func.count().filter(Report.status == Report.Status.DECLINED).label(Report.Status.DECLINED),
+                func.count().filter(Report.status == Report.Status.WAITING).label(Report.Status.WAITING),
```
Здесь статус skipped. Waiting в отчёте не нужен.

ADesBiysk commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,48 @@
+        tasks_statistic = await self.__task_repository.get_tasks_statistics_report()
+        # data = await добавляем какие-то другие данные из базы 
+        workbook = await self.__generate_task_report(tasks_statistic=tasks_statistic, workbook=workbook)
+        # await self.__generate_task_report(tasks_statistic=data, workbook=workbook)
```
Закомментированный код нужно убрать.

ADesBiysk reviewed (commented):
Обрати, пожалуйста, внимание на этот комментарий. https://github.com/Studio-Yandex-Practicum/lomaya_baryery_backend/pull/164/files#r1141394100

vatut007 committed:
Fixed remarks

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,103 @@
+        """Создаёт лист внутри отчёта."""
+        return workbook.create_sheet(sheet_name)
+
+    def add_header(self, analytic_task_report: BaseAnalyticReportSettings, worksheet: Worksheet) -> None:
```
Давай поменяем аргументы местами, поскольку воркшит главнее чем настройки

AntonZelinsky replied:
Предлагаю все методы сделать защищёнными, чтобы к ним не обращались из вне, так как это нарушит внутреннюю кухню формирования отчётов.

AntonZelinsky commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,41 @@
+                 sheet_name: str,
+                 header_data: tuple[tuple[str]],
+                 footer_data: Optional[tuple[str]] = None,
+                 row_count: Optional[int] = None):
```
А почему тут рациональный параметр, если в конкретных экземплярах этого класса ты сразу устанавливает 0? По сути, он никогда не будет таким.
Ну, и имеет смысл устанавливать это значение прямо тут, нет же кейсов, когда по дефолту идёт другое значение.

А футтера точно может не быть? Если таких кейсов нет, а кажется что не может, то там тоже не нужна опциональность

AntonZelinsky commented on src/api/routers/analytics.py:
```diff
@@ -0,0 +1,52 @@
+@cbv(router)
+class AnalyticsCBV:
+    analytics_service: AnalyticsService = Depends()
+    token: HTTPAuthorizationCredentials = Depends(HTTPBearer())
```
Приватные

AntonZelinsky commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,45 @@
+    def __init__(self, task_repository: TaskRepository = Depends()) -> None:
+        self.__task_repository = task_repository
+        self.__task_report_builder = AnalyticReportBuilder()
+        self._analytic_report_builder = AnalyticReportBuilder()
```
Если я не ошибаюсь, на 19 и 20 строке создаются од раковые экземпляры класса., так и должно быть?

И почему тут композиция, а не агрегация?

vatut007 replied:
При агрегации выходит ошибка fastapi.exceptions.FastAPIError: Invalid args for response field! Hint: check that <class 'src.excel_generator.builder.AnalyticReportBuilder'> is a valid Pydantic field type. If you are using a return type annotation that is not a valid Pydantic field (e.g. Union[Response, dict, None]) you can disable generating the response model from the type annotation with the path operation decorator parameter response_model=None. Read more: https://fastapi.tiangolo.com/tutorial/response-model/

AntonZelinsky commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,45 @@
+                                     tasks_statistic: tuple[TasksAnalyticReportDto],
+                                     workbook: Optional[Workbook] = None) -> Workbook:
+        """Генерация отчёта с заданиями."""
+        return await AnalyticReportDirector(self.__task_report_builder).generate_report(
```
И к этому директору вопрос, почему он тут создаётся, а не с помощью агрегации в конструктор инжектится?

AntonZelinsky commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,45 @@
+    async def generate_task_report(self) -> StreamingResponse:
+        """Генерация отчёта с заданиями."""
+        workbook = self.__task_report_builder.create_workbook()
+        tasks_statistic = await self.__task_repository.get_tasks_statistics_report()
+        workbook = await self.__generate_task_report(tasks_statistic=tasks_statistic, workbook=workbook)
```
Публичные функции всегда создают пустой воркбук и вызывают функцию (1 вызов, 1 статистика добавляется в книжку, то есть, эти две строки должны быть в какой-то отдельной функции) для добавления страницы со статистикой в книгу. 
Различая между полной статистикой и статистикой задач такая, что статистика задач, это одна из множества статистик полной статистики. Следовательно, приватную функцию генерации статистики задач должны вызывать и в статистике задач и в полной (в полной сделай два вызова, чтобы в книге было два одинаковых отчёта по задачам).

AntonZelinsky commented on src/excel_generator/analytic_report_director.py:
```diff
@@ -0,0 +1,32 @@
+        self,
+        data: tuple[TasksAnalyticReportDto],
+        workbook: Optional[Workbook] = None,
+        analytic_task_report_full: Optional[BaseAnalyticReportSettings] = None
```
Параметр не может быть опцаналтеым, иначе, всё сломается при обращении к её параметрам.
Вообще, везде пересмотри свою аннотацию, она много где не правильная.

AntonZelinsky commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,41 @@
+
+
+class TaskAnalyticReportSettings(BaseAnalyticReportSettings):
+    """Строитель отчёта для заданий"""
```
```suggestion
    """Конфигурация отчёта для заданий"""
```

AntonZelinsky commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,41 @@
+        self.fget = fget
+
+    def __get__(self, owner_self, owner_cls):
+        return self.fget(owner_cls)
```
А что это такое и зачем оно нам нужно?

vatut007 replied:
Я хотел использовать footer_data как атрибут класса. При использованиие propety, у меня выходила ошибка " TypeError: 'property' object is not iterable", поэо я решил сделать и сипользовать этот декоратор

vatut007 replied:
Я хотел сделать footer_data как атрибут класса, но при этом чтобы было известно зачнение row_count. Если ипользовать декаратор propety выходит ошибка TypeError: 'property' object is not iterable
При этом декатораторе такой ошибки не выходит

AntonZelinsky replied:
Ошибка потому, что что-то не так делаешь. Это какой-то очень грязный костыль. 
Погугли как циклом пройтись по списку строк. 
МОжешь накидать маленькую программку и в отладчике поковырять

AntonZelinsky commented on src/excel_generator/analytic_report_director.py:
```diff
@@ -0,0 +1,32 @@
+from src.excel_generator.task_builder import BaseAnalyticReportSettings
+
+
+class AnalyticReportDirector:
```
Что-то я перестал видеть смысл л от этого класса, кажется, что эту функцию можно вынести в билдер. давайте попробуем так сделать

vatut007 committed:
Fixed remarks after Anton

AntonZelinsky commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,46 @@
+        """Генерация отчёта с заданиями."""
+        workbook = self.__task_report_builder._create_workbook()
+        workbook = await self.__generate_task_report(workbook=workbook)
+        stream = await self.__task_report_builder._get_report_response(workbook=workbook)
```
Это же защищённый метод, а ты вызываешь его из вне... 
Это вообще законно? 
Кажется, что не должно работать, но в любом случае, так делать нельзя.

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,124 @@
+
+    async def _get_report_response(self, workbook: Workbook) -> StreamingResponse:
+        """Создание ответа."""
+        stream = await self.__save_report(workbook)
```
Неверная Аннотация. Ты же строкой выше возвращаешь BytesIO, а тут почему-то другое. Эта функция точно нужна?
Предлагаю ту что выше сделать вообще публичной и статической

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,124 @@
+from src.excel_generator.task_builder import BaseAnalyticReportSettings
+
+
+class AnalyticReportBuilder(abc.ABC):
```
Класс у тебя не абстрактный и его никто не наследует, абц лишнее. Думаю, из-за этого и не работает Depends в сервис аналитики

AntonZelinsky commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,39 @@
+    sheet_name: str = "Задачи"
+    header_data: tuple[str] = (
+        ("Задача", "Кол-во принятых отчётов", "Кол-во отклонённых отчётов", "Кол-во не предоставленных отчётов"),
+    )
```
```suggestion
    header_data: tuple[str] =
        ("Задача", "Кол-во принятых отчётов", "Кол-во отклонённых отчётов", "Кол-во не предоставленных отчётов"),
```
Писали уже сто раз, избавиться от лишней вложенности.

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,124 @@
+                cell.border = self.Styles.BORDER.value
+        worksheet.column_dimensions["A"].width = self.Styles.WIDTH.value
+
+    async def generate_report(
```
Сделай метод первым в Класе, поскольку он публичный.
Вообще, такое ощущение, что у тебя не включены различные статические анализаторы, много вопросов по стилистике коде

AntonZelinsky commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,46 @@
+                                     workbook: Workbook) -> Workbook:
+        """Генерация полного отчёта."""
+        workbook = await self.__generate_task_report(workbook=workbook)
+        workbook = await self.__generate_task_report(workbook=workbook)
```
Не нужно каждый раз переобъявлять переменную workbook, просто вызывай функцию и передавай аргумент.
Можешь вообще урать от туда возврат результата, чтобы не было соблазнов.


И объясните мне наконец, зачем нужно передавать так аргументы `workbook=workbook`? Если мне уже где-то ответили, то извиняюсь и прошу направить на тот ответ.

AntonZelinsky reviewed (approved):
@ADesBiysk @Ivan-Skvortsov Мы на финишной прямой, осталось отполировать код и можно мержить.

Только нужно пройтись по коментам которые сейчас написал и прошлый раз и проверить, что всё исправлено, а то вижу, что есть какие-то недочёты.

vatut007 committed:
fix remarks

AntonZelinsky commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,46 @@
+    async def generate_full_report(self) -> StreamingResponse:
+        """Генерация полного отчёта."""
+        workbook = self.__task_report_builder.create_workbook()
+        await self.__generate_full_report(workbook)
```
ВОзможно, этот метод нужно упростить, если внутри него идёт просто вызов генерации различных репортов и будет это происходить только один раз и для этого места, то нет особо смысла выделять для этого отдельный метод.

AntonZelinsky commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,110 @@
+            sheet_name=analytic_task_report_full.sheet_name
+        )
+        analytic_task_report_full.row_count = 0
+        self._add_header(worksheet, analytic_task_report_full)
+        self._add_data(worksheet, data, analytic_task_report_full)
+        self._add_footer(worksheet, analytic_task_report_full)
+        self._apply_styles(worksheet)
```
Все Методы делаем приватными, а не защищёнными.

AntonZelinsky commented on src/excel_generator/task_builder.py:
```diff
@@ -0,0 +1,29 @@
+class BaseAnalyticReportSettings:
+    def __init__(self,
+                 sheet_name: str,
+                 header_data: tuple[tuple[str]],
+                 row_count: int):
+        self.sheet_name = sheet_name
+        self.header_data = header_data
+        self.row_coun = row_count
```
Предлагаю сделать дата класс и сделать просто свойства с типом, а не инициализации непонятно чего через конструктор, который потом никогда больше не будет вызван.

ADesBiysk commented on src/api/routers/analytics.py:
```diff
@@ -0,0 +1,54 @@
+        Формирует отчёт с общей статистикой выполнения задач во всех сменах.
+
+        Содержит:
+        - cписок всех заданий;
```
Здесь первая c - английская.

ADesBiysk commented on src/api/routers/analytics.py:
```diff
@@ -0,0 +1,54 @@
+        self,
+    ) -> StreamingResponse:
+        """Формирует excel файл со всеми отчётами."""
+        filename = f"report_{datetime.now()}.xlsx"
```
Лучше чтобы имена файлов отличались в разных эндпойнтах. Можно просто добавить tasks и full.

ADesBiysk commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,46 @@
+        stream = await self.__task_report_builder.get_report_response(workbook)
+        return stream
+
+    async def generate_task_report(self) -> StreamingResponse:
```
Возвращаем BytesIO. И у предыдущего тоже.

ADesBiysk commented on src/core/services/analytics_service.py:
```diff
@@ -0,0 +1,46 @@
+        self.__task_report_builder = task_report_builder
+        self.__task_repository = task_repository
+
+    async def __generate_task_report(self,
```
Неправильная аннотация. Метод ничего не возвращает.

ADesBiysk commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,110 @@
+from src.excel_generator.task_builder import BaseAnalyticReportSettings
+
+
+class AnalyticReportBuilder():
```
Если не наследуемся, то скобки не нужны.

ADesBiysk commented on src/excel_generator/builder.py:
```diff
@@ -0,0 +1,110 @@
+        workbook.remove_sheet(workbook.active)
+        return workbook
+
+    def _create_sheet(self, workbook: Workbook, sheet_name: str) -> Workbook:
```
Здесь аннотация неправильная. Возвращаем Worksheet

ADesBiysk replied:
И метод можно сделать статическим.

ADesBiysk reviewed (approved):
Посмотри ещё раз аннотации типов. Есть путаница между worksheet и workbook

vatut007 committed:
Merge remote-tracking branch 'origin/develop' into feature/report_excel

vatut007 committed:
fix remarks

vatut007 committed:
fix remarks
